codec-http2/src/main/java/io/netty/handler/codec/http2/Http2HeadersValidator.java

```diff
@@ -0,0 +1,143 @@
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Connection-speficic headers like [%s] must not be used with HTTP/2.", header);
+            }
+        }
+
+        final CharSequence teHeader = headers.get(TE);
+        if (teHeader != null && !AsciiString.contentEqualsIgnoreCase(teHeader, TRAILERS)) {
+            throw streamError(streamId, PROTOCOL_ERROR,
+                              "TE header must not contain any value other than \"%s\"", TRAILERS);
+        }
+    }
+
+    /**
+     * Validates Pseudo-Header fields according to
+     * <a href="https://tools.ietf.org/html/rfc7540#section-8.1.2.3">RFC7540, section-8.1.2.3</a>
+     */
+    static void validatePseudoHeaders(Http2Headers headers, int streamId) throws Http2Exception {
+        final CharSequence method = headers.get(METHOD.value());
+        if (method == null) {
+            throw streamError(streamId, PROTOCOL_ERROR,
+                              "Mandatory header [:method] is missing.");
+        }
+
+        if (HttpMethod.CONNECT.asciiName().contentEqualsIgnoreCase(method)) {
+            if (headers.contains(SCHEME.value())) {
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Header [:scheme] must be omitted when using CONNECT method.");
+            }
+
+            if (headers.contains(PATH.value())) {
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Header [:path] must be omitted when using CONNECT method.");
+            }
+
+            if (headers.getAll(METHOD.value()).size() > 1) {
+                throw streamError(streamId, PROTOCOL_ERROR,
+                                  "Header [:method] should have a unique value.");
+            }
+        } else {
+            final CharSequence path = headers.get(PATH.value());
+            if (path != null && path.length() == 0) {
+                throw streamError(streamId, PROTOCOL_ERROR, "[:path] header cannot be empty.");
+            }
+
+            int methodHeadersCount = 0;
+            int pathHeadersCount = 0;
+            int schemeHeadersCount = 0;
+            PseudoHeaderName previousPseudoHeader = null;
+            for (Entry<CharSequence, CharSequence> entry : headers) {
+                final CharSequence key = entry.getKey();
+                if (!hasPseudoHeaderFormat(key)) {
+                    // We know that pseudo header appears first so we can stop
+                    // looking once we get to the first non pseudo headers.
+                    break;
+                }
+
+                final PseudoHeaderName pseudoHeader = PseudoHeaderName.getPseudoHeader(key);
+                if (previousPseudoHeader != null &&
+                    pseudoHeader.isRequestOnly() != previousPseudoHeader.isRequestOnly()) {
+                    throw streamError(streamId, PROTOCOL_ERROR, "Mix of request and response pseudo-headers.");
```

Comment:
Since we're only validating the request specific headers I think we can make this another `else` clause in the branches below.